### DISCUSSION TITLE
Fritzbox 6890v2 probleme mit push_firmware

### DISCUSSION THREAD

**Discussion by Himan2001** [Freetz-NG/freetz-ng#551] [Allgemein]
Meine 6890v2 ist extrem zickig und wollte einfach mit push_firmware kein initial image fressen.
( fs_linux_start existierte jungfräulich auch noch nicht ). Inital-Firmware war die 7.29

Mit folgender (alter Methode) hab ich dann das initial Image doch auf die Box bekommen:

- Bootloader manuell anhalten (ping/ftp login/logout)
- eva-tools: in.memory image erstellt
- eva_to_memory <image.in-memory> 192.168.178.1 0

Zumindest mit der "0" am Ende hats auch bei der 6890v2 endlich hingehaun - allerdings musste die Box bereits im Bootloader angehalten worden sein.

Können wir die Hardware zum push_firmware hinzufügen irgendwie ?

Ich weiss jetzt nicht genau, welchen Parametern das entspricht, die verschiedenen Optionen von push_firmware brachten
zumindest erstmal keinen Erfolg. 

Ich nehme an das der "ram" parameter hier per default bei push_firmware nicht passt für dieses Modell ?

Hier ein Auszug vom Enviroment:

```
Found AVM bootloader: AVM EVA Version 1.3292 0x0 0x46409
Environment read from device:

HWRevision            234
HWSubRevision         2
ProductID             Fritz_Box_HW234
SerialNumber          L4216xxxxxxxxxxxx
annex                 B
autoload              yes
bootloaderVersion     1.3292
country               049
firstfreeaddress      0x80000000
firmware_info         162.07.29
firmware_version      avm
flashsize             nor_size=0MB sflash_size=0KB nand_size=512MB
language              de
memsize               0x0c000000
mtd0                  0x0,0x2C00000
mtd1                  0x500000,0xD00000
mtd2                  0x0,0x100000
mtd3                  0x100000,0x500000
mtd4                  0xD00000,0x1500000
mtd5                  0x1500000,0x20000000
my_ipaddress          192.168.178.1
prompt                Eva_AVM
ptest

```

Aktuell braucht man beim push_firmware auch immer 2 Power ON/OFF der Box, beim ersten Durchgang wird das fs_linux_start nicht gefunden und ein "default" gesetzt und dann geht die Suche endlos weiter wenn man die Box nicht nochmal restartet. Beim 2. Power ON wird sowohl mit ftp als auch ncftp das Image zwar zur Box geschoben, wobei die Übertragung aber kurz vor dem Ende abbricht und die Box im Bootloader verbleibt.

**Comment by fda77** [comment]
Das Script von PP kenne ich nicht, ich weiss nicht ob es mit diesem Gerät kompatible ist

ftp-Zugriff: Das hängt von Computer, dessen Betriebsystem und Netzwerkgeräten zwischen PC und Fritzbox (Switch) ab. Wenn da eines zu langsam die Verbindung erkennt gehts nicht. Angeblich sollen Fritzboxen mit neuen Bootloadern nur Zugriff erlauben wenn man den Stromstecker reinsteckt

RAM: Mach ein Recovery der Fritzbox und schneiden den FTP-Control Kanal (tcp 20) mit. Dort steht drin wie AVM es macht und man kann den Wert hinterlegen. Dieser Wert muss nicht mit dem realen RAM übereinstimmen.

Oder du versuchst einfach plausible Werte bis es funktioniert...

**Reply by Himan2001** [reply]
Also zu eva_to_memory habe ich mal eben noch diese Info:

https://www.ip-phone-forum.de/threads/eva_to_memory-failing-for-7530.301400/#post-2303076

Ich kann mich noch daran erinnern das ich mal eine 7520 vor etwa 2 Jahren auf dem Tisch hatte und hier war das auch nur so möglich gewesen damals.

Was das Thema mit aktuellen Geräten angeht... Ja. eine direkte Verbindung habe ich nicht, da ich von einer VM über Switches hinweg
flashe. Ausschliesslich direkt verbunden mit einer Windose & AVM-Recovery rennt das bei der 6890v2.
Wenn ich ueber einen Switch gehe meldet auch das AVM-Recovery das es die Parameter nicht lesen kann, das hatte ich bislang
noch mit keiner anderen Box erlebt gehabt.

Kannst Du evtl. auch ohne Log mit dieser Info etwas anfangen ?:

https://github.com/PeterPawn/YourFritz/blob/main/eva_tools/eva_to_memory#L321

`The third parameter (which defaults to "1", while you need a "0" here) blocks the memory limitation and the upload should use all of the available 256 MB of RAM, resulting in an upper limit of 0x90000000 (the start address has to stay on 0x80000000, reverse any own changes here), as shown in your packet dump.`

So zumindest hats bei der 6890v2 funktioniert. Ich lese daraus das das Ram-Limit aus dem Enviroment nicht beachtet wird und der komplette (halbe) Speicherbereich verwendet wird (also 256MB beim STOR command?)

Was aber eine gute Idee wäre für Remote/Switch/VM-Anwender wäre das "push_firmware" etwas anzupassen, das man nicht zwangsweise eine direkte Verbindung braucht. Im Prinzip muss eigentlich nur eine Routine vorhanden sein, die den Bootloader anhält, sobald die Fritzbox pingbar ist (+ 2-4 Sekunden mehr bei der 6840v2), sich per FTP einloggt und die entsprechenden commands absetzt.

"eva_to_memory" hat im bereits manuell angehaltenen Bootloader sich zuerst das Enviroment aus der Box geholt und das in-memory image und die FTP commands entsprechend dann angepasst und den Transfer dann direkt im Anschluss (ohne einen weiteren Neustart) direkt gestartet. 

Zumindest würde ein pingen & bootloader anhalten beim "push_firmware" es erheblich einfacher machen vom Timing her wenn man
nicht direkt verbunden ist. Nebenbei gibt es zwischen "Port up" und der Erreichbarkeit des Port 21 bei der 6890v2 auch noch
eine grössere Verzögerung. Bei direkter Verbindung und "Port Up" kann es durchaus nochmal weitere 5 sekunden dauern bis auf den ersten Ping reagiert wird, und ab hier nochmal 2-4 Sekunden bis auch der Port 21 reagiert...

Was ich damit sagen will: Es braucht nur einen Kaltstart, dauerping, eine kurze Verzögerung oder Schleife bis ftp erreichbar ist, bootloader anhalten durch Login und das Enviroment auslesen und im Anschluss direkt schreiben geht in einer Session problemlos auch ohne eine direkte Verbindung zu haben zur Hardware.

**Comment by fda77** [comment]
Das Script von PP kenne ich nicht, ich weiss nicht ob es mit diesem Gerät kompatible ist!!!
Es gibt 6 Arten zu flashen bei AVM, PPs script unterstützt 1 davon, also nur manche Geräte.
Ich kann dazu nichts sagen, es gibt das ein eignes repo auf Github

> Kannst Du evtl. auch ohne Log mit dieser Info etwas anfangen ?:

Nein, mach ein tcpdump/wireshark vom recovery und wir wissen zu 100% was der korrekte Wert ist den AVM verwendet. 

> Was aber eine gute Idee wäre für Remote/Switch/VM-Anwender wäre das "push_firmware" etwas anzupassen

Wenn die Hardware zu langsam reagiert kann die Software nichts machen.
Und wenn PF zu langsam ist kannst du es ja anpassen, schau rein und such nach "ping"

**Reply by Himan2001** [reply]
... naja zumindest habe ich mal das Power-ON Verhalten der 6890v2 mit dem Enviroment "fürs Archiv" dargestellt und erläutert, wie das Timing hier abrennt - selbst wenn man lokal verbunden ist mit dem Gerät ;-)

Das das eva_to_memory "kompatibel" ist zur 6890v2 habe ich ja erläutert. Anders hab ich ja freetz nicht auf die Box erstmalig
drauf bekommen. Ums (lokale) logging kümmere ich mich noch....

**Reply by fda77** [reply]
Ich finde einen Swich dazwischen besserbesten, am unmanaged. Daher klappt bei mir die Erkennung gut
